### PR TITLE
Add support for sub images

### LINKED ISSUE #281: Add SubImage
Fonts are "just" spritemaps. Should e-g have builtin support for sprites that can handle animations and other standard sprite needs, with the fonts migrated to use that system? The alternative is to let that be built in an external crate, but I don't think there'd be that much work to adapt the existing font system to general sprite support.

### PR DESCRIPTION
**Pull request by rfuest**
Hi! Thank you for helping out with Embedded Graphics development! Please:

- [ ] Check that you've added passing tests and documentation
- [ ] Add a simulator example(s) where applicable
- [ ] Add a `CHANGELOG.md` entry in the **Unreleased** section under the appropriate heading (**Added**, **Fixed**, etc) and appropriate crate (`embedded-graphics`, `simulator`, `tinytga`, `tinybmp`) if your changes affect the **public API**
- [ ] Run `rustfmt` on the project
- [ ] Run `just build` (Linux/macOS only) and make sure it passes. If you use Windows, check that CI passes once you've opened the PR.

## PR description

This isn't quite finished yet, but I would like to get feedback on some aspects of the implementation. Some performance optimizations aren't implemented yet and the docs and examples need work. The `image-subimage.rs` example will be replaced by a simpler example that is more focused on demonstrating the feature, but the current example should be fun to play with.

To allow more flexible drawing of image data, `Image` no longer works with iterators and uses a `ImageDrawable` instead. `ImageDrawable` implementations can use all drawing methods made available by a `DrawTarget`. One example of the additional flexibility can be seen in `tinytga`, where images with top left origin are drawn using `fill_contiguous` and `draw_iter` is used as a fallback for all other origins. This could later be improved even further, e.g. by using the scanline table, that is available in some TGA files, or by issuing one `fill_contiguous` call per scanline for images with bottom left origin.

Currently the old iterator based approach is removed completely. But I'm sure that @jamwaffles won't be pleased with this :wink:. Do we need some type level guarantee that `ImageDrawable`s also provide a way to generate pixel iterators or can we make this optional? If we require it, how should we handle the different types of iterators (contiguous, pixels)?

Some naming related problems came up:

* Is there a better name for "contiguous iterator"? The name doesn't imply that its items are color values.
* Do you think that the name `SubImage` could be confusing, because it's not a part of an `Image` but a `ImageDrawable`? We could call the struct `SubImageDrawable`, but I wouldn't want to rename the method in `ImageDrawableExt` to `sub_image_drawable`. Maybe we could use `fn crop(&self, area: &Rectangle) -> SubImageDrawable<C, D>;` in `ImageDrawableExt`?

Closes #281

### CODE REVIEW

**Comment by jamwaffles** [comment]
I don't have much brainpower to look at the code properly yet, so I'll just respond to your comments:

> One example of the additional flexibility can be seen in tinytga, where images with top left origin are drawn using fill_contiguous and draw_iter is used as a fallback for all other origins.

That's really cool, nice work.

> Currently the old iterator based approach is removed completely.

This is an interesting point. Adding iterators back needs some more thought for things like contiguous, sparse, etc as you mentioned. Maybe we should tackle this later if someone has a pressing need for it? I can't see it being a very disruptive addition (famous last words...)

> Is there a better name for "contiguous iterator"? The name doesn't imply that its items are color values.

In terms of documentation, "contiguous pixel color iterator" might work, although is a bit long. Code-naming-wise, `ContiguousPixels` isn't public, so we could rename this later at our leisure. I don't have a better name suggestion at the moment unfortunately.

> Do you think that the name SubImage could be confusing, because it's not a part of an Image but a ImageDrawable?

Hm. I see what you mean, but from an end user's perspective, they'll be doing something like this:

```rust
let image: Image = Image::new(...);

let sub: SubImage = image.sub_image(...);
```

I think the symmetry between `image: Image` and `sub: SubImage` make sense in terms of naming here. The `ImageDrawable` is just an implementation detail here, no?

**Comment by rfuest** [comment]
> Hm. I see what you mean, but from an end user's perspective, they'll be doing something like this: ...
 
They wouldn't, `sub_image` needs to be called on an `ImageDrawable` and not `Image`, because the underlying image data should be cropped and not the version that is shifted by some offset:
```rust
let tga = Tga::new(...);
let sub_image_drawable = tga.sub_image(...);

let image: Image<_, Rgb888> = Image::new(tga, Point::new(1, 2));
image.draw(&mut display)?;

let sub_image: Image<_, Rgb888> = Image::new(sub_image_drawable, Point::new(10, 20));
sub_image.draw(&mut display)?;
```

But this code example shows another Problem: The `Tga` object doesn't define the color types which makes it necessary to specify the color type each time it or one of its sub images is used.

**Comment by jamwaffles** [comment]
Oh ok my bad. I didn't look at the demo code closely enough. Could the `sub_image` method still be moved to `Image` though? That would mean that the source BMP/TGA/etc is "in e-g land" once it's wrapped in an `Image`, i.e. the raw image doesn't need to be referenced anymore.

`Image::sub_image` could return a sub image relative from the top left corner of the parent image, regardless of translation, but have its own top left set to `(0, 0)`. Does that make sense to implement? e.g.

```rust
// For example, 64x64px
let tga = Tga::from_slice(PIPES_IMAGE).unwrap();

// Base image offset by 10, 20
let image = Imge::new(tga, Point::new(10, 20));

// Sub image with top-left at (0, 0), picking an 8x8px square offset by 8x8px from the image's 
// top left corner.
let sub = image.sub_image(Rectangle::new(Point::new(8, 8), Point::new(8, 8)));
```

`sub` could be moved around with `.translate()` calls, or `Image::sub_image` could take another offset argument, much like we do with `Text`.

**Comment by rfuest** [comment]
What type would `Image::sub_image` return in your example?

I think we should keep `Image` and `ImageDrawable` separate from each other. Having a `sub_image` method on `Image` breaks this distinction and I don't like that some methods would use the position of the `Image` while other would ignore it.

One solution I considered is a `ImageFile` struct:
```rust
// The image file format is determined by setting an explicit type
let tga: ImageFile<Tga, Rgb565> = ImageFile::from_slice(PIPES_IMAGE).unwrap();

//no need to specify the color type, because it's already defined in `tga`
let sub = tga.sub_image(...);

let i1 = Image::new(tga, p1);
let i2 = Image::new(sub, p2);
```
This way the object that is created for the image file would already contain the color type information, which isn't the case at the moment. It would provide a generic interface for all image file types, but perhaps the API is slightly unusual because you need to specify an explicit type to set the file format, although this is already required for the color type anyway.

**Comment by jamwaffles** [comment]
> One solution I considered is a ImageFile struct:

That looks like it could be a good solution. I think it can still be called `Image` though as it's one less naming change for 0.7 (even though the type params would change). If you'd like to rename it to `ImageFile` I'm good with that.

Presumably `ImageRaw` is also usable where `Tga` is above?

> perhaps the API is slightly unusual because you need to specify an explicit type to set the file format

I think it's ok tbh. Better to be explicit :)

**Comment by rfuest** [comment]
> I think it can still be called `Image` though as it's one less naming change for 0.7 (even though the type params would change). If you'd like to rename it to `ImageFile` I'm good with that.

`ImageFile` is an implementation of `ImageDrawable` and wasn't supposed to replace `Image`. `Image` would still be required to position the `ImageDrawable` and convert it into a `Drawable`.

> Presumably `ImageRaw` is also usable where `Tga` is above?

No, I don't think that it would. `ImageFile` is supposed to provide an easy interface for all image formats that contain a header and don't require additional information. To create a `ImageRaw` instance you need to provide the dimensions in addition to the image data. But because `ImageRaw` already sets the color type there is no need to warp it in `ImageFile` anyway.

**Comment by rfuest** [comment]
I've pushed a first draft of the `ImageFile` type to this PR, which should make it easier to discuss the API.
```rust
let image = Bmp::from_slice(include_bytes!("./colors_grey8.bmp")).unwrap();
let image: Image<_, Gray8> = Image::new(&image, Point::zero());
```
got replaced by
```rust
let image: ImageFile<Bmp, Gray8> = ImageFile::from_slice(include_bytes!("./colors_grey8.bmp")).unwrap();
let image = Image::new(&image, Point::zero());
```
Being able to specify the image format and the color type in a single line makes this more readable IMO.

**Comment by rfuest** [comment]
I've changed my mind and removed `ImageFile`. Instead I've added `EgBmp` and `EgTga` in the tiny... libraries. The purpose of these types is to set the right color type and they must now be used instead of the raw `Bmp` and `Tga` types:
```rust
// before:
let image = Bmp::from_slice(include_bytes!("./colors_grey8.bmp")).unwrap();
let image: Image<_, Gray8> = Image::new(&image, Point::zero());

// after:
let image: EgBmp<Gray8> = EgBmp::from_slice(include_bytes!("./colors_grey8.bmp")).unwrap();
let image = Image::new(&image, Point::zero());
```
This change made it possible to use an associated color type for `ImageDrawable`, which allowed me to clean up the types in the `image` module.

**Comment by rfuest** [comment]
This should be ready for review now.

Because IMO `tinybmp` and `tinytga` are primarily used by e-g users I decided to rename `Bmp` and `Tga` to `BmpRaw` and `TgaRaw` and reuse the original names for the new embedded-graphics types, instead of adding an `Eg` prefix.

**Comment by jamwaffles** [comment]
> Because IMO tinybmp and tinytga are primarily used by e-g users I decided to rename Bmp and Tga to BmpRaw and TgaRaw and reuse the original names for the new embedded-graphics types, instead of adding an Eg prefix.

Ok, makes sense to me. More naming bikeshed: would `TgaData` and `BmpData` be better names? If you disagree, let's leave them as `*Raw` as it doesn't make too much difference.

> tinybmp and tinytga are primarily used by e-g users

Should we enable the e-g feature by default on both crates do you think?

File: tinybmp/CHANGELOG.md [hunk 1]
```diff
@@ -16,11 +20,11 @@
 
 ## [0.2.1] - 2020-02-17
 
-- #244 Added `.into_iter()` support to the `Bmp` struct to get an iterator over every pixel in the image.
+- [#244](https://github.com/jamwaffles/embedded-graphics/pull/244) Added `.into_iter()` support to the `Bmp` struct to get an iterator over every pixel in the image.
 
 ### Changed
 
-- **(breaking)** #247 "reverse" integration of tinybmp into [`embedded-graphics`](https://crates.io/crates/embedded-graphics). tinybmp now has a `graphics` feature that must be turned on to enable embedded-graphics support. The `bmp` feature from embedded-graphics is removed.
+- **(breaking)** [#247](https://github.com/jamwaffles/embedded-graphics/pull/247) "reverse" integration of tinybmp into [`embedded-graphics`](https://crates.io/crates/embedded-graphics). tinybmp now has a `graphics` feature that must be turned on to enable embedded-graphics support. The `bmp` feature from embedded-graphics is removed.
```
Comment (jamwaffles) [inline tinybmp/CHANGELOG.md:27]: Thanks for adding these links.

File: tinytga/src/embedded_graphics.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    C: PixelColor + From<<C as PixelColor>::Raw>,
+{
+    type Color = C;
+
+    fn draw<D>(&self, target: &mut D) -> Result<(), D::Error>
+    where
+        D: DrawTarget<Color = C>,
+    {
+        // TGA files with the origin in the top left corner can be drawn using `fill_contiguous`.
+        // All other origins are drawn by falling back to `draw_iter`.
```
Comment (jamwaffles) [inline tinytga/src/embedded_graphics.rs:33]: Would this be worth mentioning in a `# Performance` section in the `Tga` struct docs?

File: embedded-graphics/src/image/mod.rs [hunk 3]
```diff
@@ -1,19 +1,19 @@
...
+//! The two main types used to draw images are [`ImageDrawable`] and [`Image`].
 //!
-//! The [`Image`] struct is a wrapper around items that implement both [`ImageDimensions`] and
-//! [`IntoPixelIter`] and allows them to be drawn to a [`DrawTarget`], reading pixel values from the
-//! implementation of [`IntoPixelIter`].
+//! [`ImageDrawable`] is implemented to add support for different image formats. This crate includes
+//! an implementation for [raw pixel data]. Additional implementations for other image formats are
+//! provided by external crates like [tinybmp] and [tinytga].
+//!
+//! The [`Image`] object is used to specify the location at which an [`ImageDrawable`] is drawn.
```
[the remark this answers is not in the document]
Comment (rfuest) [reply embedded-graphics/src/image/mod.rs:9]: Yes, it is.

File: embedded-graphics/src/iterator/contiguous.rs [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    I::Item: PixelColor,
+{
+    type Item = Pixel<I::Item>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iter.next().map(|(p, c)| Pixel(p, c))
+    }
+}
+
+// TODO: should this be made public and be added to ContiguousIteratorExt?
```
[the remark this answers is not in the document]
Comment (rfuest) [reply embedded-graphics/src/iterator/contiguous.rs]: I hadn't noticed the remaining TODO comments.

File: tinytga/src/embedded_graphics.rs (hunk 2 above)
Comment (rfuest) [reply tinytga/src/embedded_graphics.rs:41]: That's a good idea! I've added a performance section to the docs.

File: embedded-graphics/src/iterator/pixel.rs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            .map(|Pixel(p, c)| Pixel(p + self.offset, c))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{iterator::PixelIteratorExt, pixelcolor::BinaryColor};
+
+    // TODO: Add draw test
```
[the remark this answers is not in the document]
Comment (rfuest) [reply embedded-graphics/src/iterator/pixel.rs]: Thanks for the reminder, I've added the test in another location.

File: embedded-graphics/src/image/sub_image.rs [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    primitives::Rectangle,
+    transform::Transform,
+};
+
+/// Sub image.
+///
+/// A sub image is rectangular subsection of an [`ImageDrawable`]. It can, for example, be used to
+/// draw individual sprites from a larger sprite sheet.
+///
+/// To create a sub image call the [`sub_image`] method on the parent [`ImageDrawable`].
```
[the remark this answers is not in the document]
Comment (rfuest) [reply embedded-graphics/src/image/sub_image.rs]: Yes it needs one, but I didn't want to add the same example a third time.

Would a link be enough? I've now added one the the module-level docs.

File: simulator/src/bin/generate-drawing-examples.rs [hunk 7]
```diff
@@ -293,11 +293,11 @@ object with embedded-graphics.",
...
-            let tga = Tga::from_slice(
+            let tga: Tga<Rgb888> = Tga::from_slice(
                 include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/../simulator/examples/assets/rust-pride.tga"))
             ).unwrap();
             {}
-            let image: Image<Tga, Rgb888> = Image::new(&tga, Point::zero());
+            let image = Image::new(&tga, Point::zero());
             {}
             // Display the image
             image.draw(&mut display)?;
```
[the remark this answers is not in the document]
Comment (rfuest) [reply simulator/src/bin/generate-drawing-examples.rs:303]: I wouldn't add it as a separate example.

We could change the example to be less focused on TGA images and add a note about the sub image feature in the description.

File: embedded-graphics/src/image/sub_image.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (jamwaffles) [reply embedded-graphics/src/image/sub_image.rs]: Yeah that's fine. Even just a link to an example should help with discoverability.

File: simulator/src/bin/generate-drawing-examples.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (jamwaffles) [reply simulator/src/bin/generate-drawing-examples.rs:303]: Let's leave it as is. No need to clutter up the example here I think.

**Comment by rfuest** [comment]
> More naming bikeshed: would `TgaData` and `BmpData` be better names? If you disagree, let's leave them as `*Raw` as it doesn't make too much difference.

I agree that `TgaRaw` is not the greatest name, but I don't like `TgaData` either, because IMO the name suggests that the object only contains the pixel data of the TGA file and not also the header.

> Should we enable the e-g feature by default on both crates do you think?

I would support this idea. I did even consider to remove the flag entirely. The compile time and size overhead of embedded-graphics shouldn't be to great if the `tiny...` crates are used without using `embedded-graphics` to draw the image files. And the `Point` and color types would provide a nicer interface.

**Comment by jamwaffles** [comment]
> I agree that TgaRaw is not the greatest name, but I don't like TgaData either, because IMO the name suggests that the object only contains the pixel data of the TGA file and not also the header.

Good point. Let's leave it as is.

> I would support this idea. I did even consider to remove the flag entirely.

Can you open an issue to do this in another PR? If we remove it, I think we should still provide a non-e-g interface for users who want to work with the image data directly.

**Comment by rfuest** [comment]
> Can you open an issue to do this in another PR? If we remove it, I think we should still provide a non-e-g interface for users who want to work with the image data directly.

Done.

**Comment by jamwaffles** [comment]
Sweet, thanks

**Pull request merged**